pkgs/applications/networking/cluster/hadoop/default.nix

```diff
@@ -1,51 +1,145 @@
-{ stdenv, fetchurl, makeWrapper, which, jre, bash }:
+{ stdenv, fetchurl, makeWrapper, pkgconfig, which, maven, cmake, jre, bash, coreutils, glibc, protobuf2_5, fuse, snappy, zlib, bzip2, openssl }:
 
-stdenv.mkDerivation rec {
 
-  name = "hadoop-2.2.0";
+let
+  common = { version, sha256, dependencies-sha256, tomcat }:
+    let
+      # compile the hadoop tarball from sources, it requires some patches
+      binary-distributon = stdenv.mkDerivation rec {
+        name = "hadoop-${version}-bin";
+        src = fetchurl {
+          url = "mirror://apache/hadoop/common/hadoop-${version}/hadoop-${version}-src.tar.gz";
+          inherit sha256;
+        };
 
-  src = fetchurl {
-    url = "mirror://apache/hadoop/common/${name}/${name}.tar.gz";
-    sha256 = "0r0kx8arsrvmcfy0693hpv4cz3i0razvk1xa3yhlf3ybb80a8106";
+        # perform fake build to make a fixed-output derivation of dependencies downloaded from maven central (~100Mb in ~3000 files)
+        fetched-maven-deps = stdenv.mkDerivation {
+          name = "hadoop-${version}-maven-deps";
+          inherit src nativeBuildInputs buildInputs configurePhase mavenFlags;
+          buildPhase = ''
```

Comment:
Have you seen https://github.com/NixOS/nixpkgs/pull/32695/commits/0cfda0b162f61a561ae83824cf53b61d68b21f4a?

Reply:
No, I haven't.
This could be eventually combined, but so far 0cfda0b lacks some features:
1. no way to pass ```mavenFlags```
2. no watchdog, Maven really hangs building big projects and it happens very often.
3. there are more ephemeral file names than ```_remote.repositories``` and ```*.lastUpdated``` so I ended up in whitelist

PS. actually I am quite pessimistic about generic functions for fetching artifacts and ```maven```/```sbt``` building. There are 3 or 4 of them in ```nixpkgs``` not to mention many private solutions and their reusability is quite poor (for example, presenting ```cmake```-subprojects in hadoop made it tricky, and almost every big project have something weird like that)

Reply:
@yegortimoshenko 
among ephemeral files: there are also ```resolver-status.properties```, you might want to add it to 0cfda0b
among stable files: there are also ```*.{dll,so,dylib}```

Reply:
Thanks! I'm not sure if blacklist or whitelist is better.

Reply:
>  I'm not sure if blacklist or whitelist is better.

Build twice and compare outputs :)